From f748bf5df63f5cbe502161c65d1149eaf6674b06 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:00:46 -0400
Subject: [fs] btrfs: fix null pointer deference at btrfs_sysfs_add_one+0x105

Message-id: <1407179109-13714-160-git-send-email-sandeen@redhat.com>
Patchwork-id: 87192
O-Subject: [RHEL7.1 PATCH 159/422] btrfs: fix null pointer deference at btrfs_sysfs_add_one+0x105
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Anand Jain <Anand.Jain@oracle.com>

commit f085381e6d08f4c8d6882825f31accd455c54d70
Author: Anand Jain <Anand.Jain@oracle.com>
Date:   Wed Jan 15 17:22:28 2014 +0800

    btrfs: fix null pointer deference at btrfs_sysfs_add_one+0x105

    bdev is null when disk has disappeared and mounted with
    the degrade option

    stack trace
    ---------
    btrfs_sysfs_add_one+0x105/0x1c0 [btrfs]
    open_ctree+0x15f3/0x1fe0 [btrfs]
    btrfs_mount+0x5db/0x790 [btrfs]
    ? alloc_pages_current+0xa4/0x160
    mount_fs+0x34/0x1b0
    vfs_kern_mount+0x62/0xf0
    do_mount+0x22e/0xa80
    ? __get_free_pages+0x9/0x40
    ? copy_mount_options+0x31/0x170
    SyS_mount+0x7e/0xc0
    system_call_fastpath+0x16/0x1b
    ---------

    reproducer:
    -------
    mkfs.btrfs -draid1 -mraid1 /dev/sdc /dev/sdd
    (detach a disk)
    devmgt detach /dev/sdc [1]
    mount -o degrade /dev/sdd /btrfs
    -------

    [1] github.com/anajain/devmgt.git

    Signed-off-by: Anand Jain <Anand.Jain@oracle.com>
    Tested-by: Hidetoshi Seto <seto.hidetoshi@jp.fujitsu.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/sysfs.c b/fs/btrfs/sysfs.c
index 782374d..865f4cf 100644
--- a/fs/btrfs/sysfs.c
+++ b/fs/btrfs/sysfs.c
@@ -578,8 +578,14 @@ static int add_device_membership(struct btrfs_fs_info *fs_info)
   return -ENOMEM;
 
  list_for_each_entry(dev, &fs_devices->devices, dev_list) {
-  struct hd_struct *disk = dev->bdev->bd_part;
-  struct kobject *disk_kobj = &part_to_dev(disk)->kobj;
+  struct hd_struct *disk;
+  struct kobject *disk_kobj;
+
+  if (!dev->bdev)
+   continue;
+
+  disk = dev->bdev->bd_part;
+  disk_kobj = &part_to_dev(disk)->kobj;
 
   error = sysfs_create_link(fs_info->device_dir_kobj,
        disk_kobj, disk_kobj->name);
-- 
1.7.1